crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs

```diff
@@ -127,31 +129,164 @@ impl Extension for QueryLimitsChecker {
+                        if f.node.name.node == "executeTransactionBlock"
+                            || f.node.name.node == "dryRunTransactionBlock"
+                        {
+                            // we know this has always two arguments as per the schema
+                            let tx_bytes_arg = &f.node.arguments[0].1.node;
+                            let sigs_arg = &f.node.arguments[1].1.node;
+                            let tx_bytes_len = get_value_str_len(tx_bytes_arg, variables);
+                            let sigs_len = get_value_str_len(sigs_arg, variables);
+
+                            // Check it the txBytes is larger than the allowed limit
+                            if tx_bytes_len > cfg.limits.max_mutation_payload_size as usize {
+                                metrics
+                                    .request_metrics
+                                    .query_payload_too_large_size
+                                    .observe(tx_bytes_len as f64);
+                                info!(
+                                    query_id = %query_id,
+                                    session_id = %session_id,
+                                    error_cod_ = code::BAD_USER_INPUT,
+                                    "Mutation payload txBytes size is too large: {}",
+                                    tx_bytes_len
+                                );
+                                return Err(graphql_error_at_pos(
+                                    code::BAD_USER_INPUT,
+                                    format!(
+                                        "Mutation payload txBytes size of {} node is too large. The maximum allowed is {} bytes",
+                                        f.node.name.node,
+                                        cfg.limits.max_mutation_payload_size
+                                    ),
+                                    f.pos
+                                ));
+                            }
+                            tx_bytes_sigs.insert(f.pos, (tx_bytes_len, sigs_len));
+                        }
+                    }
+                }
+            }
+
+            // Case 2: check if the read part of the query (so query minus the txBytes and signatures) is larger
+            // than the max query payload size. Given that the query.len() does not include the
+            // resolved variables, it is totally OK to use it here
+            // However, if variables are used, it can be that the txBytes + signatures are larger
+            // than the whole query, so we need to check that as well
+            let all_tx_bytes = tx_bytes_sigs
+                .iter()
+                .fold(0, |acc, (_, (tx_bytes, _))| acc + tx_bytes);
+            let all_sigs_bytes = tx_bytes_sigs
+                .iter()
+                .fold(0, |acc, (_, (_, sigs_bytes))| acc + sigs_bytes);
+
+            // in this case, variables are most likely not used (or their size is small)
```

Comment:
I don't think this is a reasonable assumption, especially in an adversarial setting.

Reply:
I'll think more about it, but the issue is that we cannot subtract a `query.len()` when variables are used for passing `txBytes`, as `txBytes` could be a lot bigger than `query.len()` and this will err. Probably need to do an extra check here.

Reply:
I fixed this by basically thinking like this:
- query does not have the resolved variables. 
- the variables we are interested in are: txBytes + signatures for mutation, and txBytes for dryRun. 
- we can use `variables.to_string().len()` to get the length of the variables, then remove the txBytes + signatures (or just txBytes for dryRun), and add query length to determine if the "read" part of a request (mutation or query) 
is within limits.